File: resources/worker_sdk/node/callit/kv.js
const http = require("node:http");
const https = require("node:https");
const { URL } = require("node:url");

async function request(action, payload) {
  const baseURL = (process.env.CALLIT_MAGIC_API_BASE_URL || "").replace(/\/$/, "");
  const workerID = process.env.CALLIT_WORKER_ID || "";
  const requestID = process.env.CALLIT_REQUEST_ID || "";
  if (!baseURL || !workerID || !requestID) {
    throw new Error("当前环境缺少 kv 所需上下文");
  }

  const requestURL = new URL(`${baseURL}/kv/${action}`);
  const body = JSON.stringify(payload);
  const transport = requestURL.protocol === "https:" ? https : http;

  const response = await new Promise((resolve, reject) => {
    const req = transport.request({
      protocol: requestURL.protocol,
      hostname: requestURL.hostname,
      port: requestURL.port,
      path: `${requestURL.pathname}${requestURL.search}`,
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "Content-Length": Buffer.byteLength(body),
        "X-Callit-Worker-Id": workerID,
        "X-Callit-Request-Id": requestID,
      },
    }, (res) => {
      const chunks = [];
      res.on("data", (chunk) => chunks.push(chunk));
      res.on("end", () => {
        resolve({
          ok: (res.statusCode || 0) >= 200 && (res.statusCode || 0) < 300,
          status: res.statusCode || 0,
          text: Buffer.concat(chunks).toString("utf-8"),
        });
      });
    });

    req.on("error", (error) => {
      reject(new Error(`请求 kv 服务失败: ${error.message}`));
    });
    req.setTimeout(5000, () => {
      req.destroy(new Error("请求 kv 服务超时"));
    });
    req.write(body);
    req.end();
  });

  const text = response.text.trim();
  if (!response.ok) {
    let message = text || `kv 请求失败: ${response.status}`;
    try {
      const parsed = text ? JSON.parse(text) : {};
      if (parsed && parsed.error) {
        message = parsed.error;
      }
    } catch (_) {
      // ignore
    }
    throw new Error(message);
  }
  return text ? JSON.parse(text) : null;
}

function normalizeNamespace(namespace) {
  const workerID = (process.env.CALLIT_WORKER_ID || "").trim();
  if (!workerID) {
    throw new Error("当前环境缺少 CALLIT_WORKER_ID");
  }
  if (namespace === null || namespace === undefined) {
    return workerID;
  }
  if (typeof namespace !== "string") {
    throw new TypeError("kv.newClient 的 namespace 必须是 string、null 或 undefined");
  }
  const normalizedNamespace = namespace.trim();
  if (!normalizedNamespace) {
    return workerID;
  }
  return normalizedNamespace;
}

function createKVClient(namespace) {
  const normalizedNamespace = normalizeNamespace(namespace);

  return {
    async set(key, value, seconds) {
      if (typeof value !== "string") {
        throw new TypeError("kv client 的 value 必须是 string");
      }
      await request("set", {
        namespace: normalizedNamespace,
        key,
        value,
        seconds,
      });
    },
    async get(key) {
      const response = (await request("get", { namespace: normalizedNamespace, key })) || {};
      return response.value;
    },
    async delete(key) {
      await request("delete", { namespace: normalizedNamespace, key });
    },
    async increment(key, step = 1) {
      const response = (await request("increment", { namespace: normalizedNamespace, key, step })) || {};
      return Number.parseInt(String(response.int_value || 0), 10);
    },
    async expire(key, seconds) {
      await request("expire", { namespace: normalizedNamespace, key, seconds });
    },
    async ttl(key) {
      const response = (await request("ttl", { namespace: normalizedNamespace, key })) || {};
      return Number.parseInt(String(response.seconds || 0), 10);
    },
    async has(key) {
      const response = (await request("has", { namespace: normalizedNamespace, key })) || {};
      return Boolean(response.exists);
    },
  };
}

const kv = {
  newClient(namespace) {
    return createKVClient(namespace);
  },
};

module.exports = {
  kv,
  normalizeNamespace,
};

File: resources/worker_sdk/python/callit/kv.py
import json
import os
import urllib.error
import urllib.request


def _request(action, payload):
    base_url = (os.environ.get("CALLIT_MAGIC_API_BASE_URL") or "").rstrip("/")
    worker_id = os.environ.get("CALLIT_WORKER_ID") or ""
    request_id = os.environ.get("CALLIT_REQUEST_ID") or ""
    if not base_url or not worker_id or not request_id:
        raise RuntimeError("当前环境缺少 kv 所需上下文")

    request = urllib.request.Request(
        url=f"{base_url}/kv/{action}",
        data=json.dumps(payload, ensure_ascii=False).encode("utf-8"),
        headers={
            "Content-Type": "application/json",
            "X-Callit-Worker-Id": worker_id,
            "X-Callit-Request-Id": request_id,
        },
        method="POST",
    )
    try:
        with urllib.request.urlopen(request) as response:
            raw = response.read().decode("utf-8").strip()
            return json.loads(raw) if raw else None
    except urllib.error.HTTPError as error:
        body = error.read().decode("utf-8", errors="ignore").strip()
        try:
            parsed = json.loads(body) if body else {}
        except json.JSONDecodeError:
            parsed = {}
        message = parsed.get("error") or body or f"kv 请求失败: {error.code}"
        raise RuntimeError(message)
    except urllib.error.URLError as error:
        raise RuntimeError(f"请求 kv 服务失败: {error}") from error


def normalize_namespace(namespace):
    worker_id = (os.environ.get("CALLIT_WORKER_ID") or "").strip()
    if not worker_id:
        raise RuntimeError("当前环境缺少 CALLIT_WORKER_ID")
    if namespace is None:
        return worker_id
    if not isinstance(namespace, str):
        raise TypeError("kv.new_client 的 namespace 必须是 string 或 None")
    normalized_namespace = namespace.strip()
    if not normalized_namespace:
        return worker_id
    return normalized_namespace


class _KVClient:
    def __init__(self, namespace=None):
        self._namespace = normalize_namespace(namespace)

    def set(self, key, value, seconds):
        if not isinstance(value, str):
            raise TypeError("kv client 的 value 必须是 string")
        _request("set", {
            "namespace": self._namespace,
            "key": key,
            "value": value,
            "seconds": seconds,
        })

    def get(self, key):
        response = _request("get", {"namespace": self._namespace, "key": key}) or {}
        return response.get("value")

    def delete(self, key):
        _request("delete", {"namespace": self._namespace, "key": key})

    def increment(self, key, step=1):
        response = _request("increment", {"namespace": self._namespace, "key": key, "step": step}) or {}
        return int(response.get("int_value", 0))

    def expire(self, key, seconds):
        _request("expire", {"namespace": self._namespace, "key": key, "seconds": seconds})

    def ttl(self, key):
        response = _request("ttl", {"namespace": self._namespace, "key": key}) or {}
        return int(response.get("seconds", 0))

    def has(self, key):
        response = _request("has", {"namespace": self._namespace, "key": key}) or {}
        return bool(response.get("exists"))


class _KVFactory:
    def new_client(self, namespace=None):
        return _KVClient(namespace)

    def newClient(self, namespace=None):
        return self.new_client(namespace)


kv = _KVFactory()
